pkg/auction/auctioneer.go

```diff
@@ -0,0 +1,88 @@
+package auction
+
+import (
+	"sync"
+
+	"github.com/lthibault/log"
+
+	"github.com/blocknative/dreamboat/pkg/structs"
+	"github.com/flashbots/go-boost-utils/types"
+)
+
+type Auctioneer struct {
+	l log.Logger
+
+	auctions [2]*Auction
+}
+
+type Auction struct {
+	mu                   sync.RWMutex
+	maxProfit            *structs.CompleteBlockstruct
+	latestBlockByBuilder map[types.PublicKey]*structs.CompleteBlockstruct
+}
+
+func NewAuctioneer(l log.Logger) *Auctioneer {
+	return &Auctioneer{
+		l: l.WithField("relay-service", "auctioneer"),
+		auctions: [2]*Auction{
+			{latestBlockByBuilder: make(map[types.PublicKey]*structs.CompleteBlockstruct)}, // slot
+			{latestBlockByBuilder: make(map[types.PublicKey]*structs.CompleteBlockstruct)}, // slot + 1
+		},
+	}
+}
+
+func (a *Auctioneer) AddBlock(block *structs.CompleteBlockstruct) {
+	auction := a.auctions[block.Header.Trace.Slot%2]
+
+	auction.mu.Lock()
+	defer auction.mu.Unlock()
+
+	auction.latestBlockByBuilder[block.Payload.Trace.Message.BuilderPubkey] = block
+
+	// always set new value and bigger slot
+	if auction.maxProfit == nil || auction.maxProfit.Header.Trace.Slot < block.Header.Trace.Slot {
+		auction.maxProfit = block
+		a.l.WithField("slot", block.Header.Trace.Slot).WithField("value", block.Header.Trace.Value.String()).Trace("new max bid")
```

Comment:
Do we really need that at prod?

Reply:
Mmm good question. This helps to keep track of the auctions, but I think you are right, the logs we have on the block submissions may be enough? In any case, having it as trace log may be useful?